src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java

```diff
@@ -98,57 +136,301 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
+                        criteria.getOid(),
+                        "",
+                        criteria.getDescription().split(":")[0]))
+            .collect(Collectors.toSet());
+    List<HumanReadableValuesetModel> valuesetList = new ArrayList<>(valuesets);
+    valuesetList.sort(
+        Comparator.comparing(HumanReadableValuesetModel::getDataCriteriaDisplay, collator));
+    return valuesetList;
+  }
+
+  boolean findUsedValueset(Measure measure, String accessToken, String id) {
+    List<String> usedValuesets =
+        dataCriteriaService.getUsedValuesets(measure.getCql(), accessToken);
+    return usedValuesets != null && !usedValuesets.isEmpty() && usedValuesets.contains(id);
+  }
+
+  List<HumanReadableCodeModel> buildCodeDataCriteriaList(List<CQLCode> cqlCodes) {
+    if (!CollectionUtils.isEmpty(cqlCodes)) {
+      List<HumanReadableCodeModel> codeList =
+          cqlCodes.stream()
+              .map(
+                  cqlCode ->
+                      HumanReadableCodeModel.builder()
+                          .name(cqlCode.getCodeName())
+                          .oid(cqlCode.getId())
+                          .codesystemName(cqlCode.getCodeSystemName())
+                          .codesystemVersion(cqlCode.getCodeSystemVersion())
+                          .isCodesystemVersionIncluded(cqlCode.isIsCodeSystemVersionIncluded())
+                          .datatype(cqlCode.getCodeIdentifier()) // ???
+                          .build())
+              .collect(Collectors.toList());
+      codeList.sort(Comparator.comparing(HumanReadableCodeModel::getDataCriteriaDisplay, collator));
+      return codeList;
+    }
+    return new ArrayList<>();
+  }
+
+  List<HumanReadableTerminologyModel> buildValuesetTerminologyList(
+      List<HumanReadableValuesetModel> valuesetModel,
+      Measure measure,
+      String accessToken,
+      List<SourceDataCriteria> sourceDataCriteria) {
+    List<HumanReadableTerminologyModel> valuesetList = new ArrayList<>(valuesetModel);
+
+    Set<HumanReadableValuesetModel> usedValueSets =
+        findUsedCQLValueSet(measure, accessToken, sourceDataCriteria);
+    if (!usedValueSets.isEmpty()) {
+      valuesetList.addAll(usedValueSets);
+    }
+
+    valuesetList.sort(
+        Comparator.comparing(HumanReadableTerminologyModel::getTerminologyDisplay, collator));
+    return valuesetList;
+  }
+
+  Set<HumanReadableValuesetModel> findUsedCQLValueSet(
+      Measure measure, String accessToken, List<SourceDataCriteria> sourceDataCriteria) {
+    List<CQLValueSet> usedValuesets =
+        dataCriteriaService.getUsedCQLValuesets(measure.getCql(), accessToken);
+    List<CQLValueSet> otherValueSets = new ArrayList<>();
+    if (!CollectionUtils.isEmpty(usedValuesets)) {
+      for (CQLValueSet cqlValueSet : usedValuesets) {
+        if (findNotExistInSourceDataCriteria(cqlValueSet.getOid(), sourceDataCriteria) == null) {
+          otherValueSets.add(cqlValueSet);
+        }
+      }
+    }
```

Comment:
I tried to match the intent using stream filter. If it looks good, then `findNotExistInSourceDataCriteria` can be removed.
```suggestion
    List<CQLValueSet> otherValueSets =
        usedValuesets.stream()
            .filter(
                vs ->
                    sourceDataCriteria.stream()
                        .noneMatch(sdc -> sdc.getOid().equalsIgnoreCase(vs.getOid())))
            .toList();
```

Reply:
Changed